File: structural/composite/file_system/python/file_system.py
import abc
from typing import List

# --- Component Interface ---
class FileSystemComponent(abc.ABC):
    """The common interface for both leaves (files) and composites (directories)."""

    def __init__(self, name: str):
        self._name = name

    def get_name(self) -> str:
        return self._name

    @abc.abstractmethod
    def get_size(self) -> int:
        """Returns the size of the component in bytes."""
        pass

    @abc.abstractmethod
    def display(self, indent: str = "") -> None:
        """Displays the component's structure."""
        pass

    # Optional: Methods for managing children (might raise errors in Leaf)
    def add(self, component: 'FileSystemComponent') -> None:
        raise NotImplementedError("Cannot add to this component")

    def remove(self, component: 'FileSystemComponent') -> None:
        raise NotImplementedError("Cannot remove from this component")

    def get_child(self, index: int) -> 'FileSystemComponent':
        raise NotImplementedError("Cannot get child from this component")


# --- Leaf Class ---
class File(FileSystemComponent):
    """Represents a leaf object (a file) in the composition."""

    def __init__(self, name: str, size: int):
        super().__init__(name)
        self._size = size

    def get_size(self) -> int:
        return self._size

    def display(self, indent: str = "") -> None:
        print(f"{indent}- {self.get_name()} ({self.get_size()} bytes)")


# --- Composite Class ---
class Directory(FileSystemComponent):
    """Represents a composite object (a directory) that can contain other components."""

    def __init__(self, name: str):
        super().__init__(name)
        self._children: List[FileSystemComponent] = []

    def add(self, component: FileSystemComponent) -> None:
        """Adds a child component (file or subdirectory)."""
        self._children.append(component)

    def remove(self, component: FileSystemComponent) -> None:
        """Removes a child component."""
        self._children.remove(component)

    def get_child(self, index: int) -> FileSystemComponent:
        """Gets a specific child component."""
        return self._children[index]

    def get_size(self) -> int:
        """Calculates the total size by summing the sizes of all children."""
        total_size = 0
        for child in self._children:
            total_size += child.get_size()
        return total_size

    def display(self, indent: str = "") -> None:
        """Displays the directory and recursively displays its children."""
        print(f"{indent}+ {self.get_name()} ({self.get_size()} bytes total)")
        for child in self._children:
            child.display(indent + "  ")

File: structural/composite/file_system/typescript/file_system.ts
// --- Component Interface (using an abstract class) ---
export abstract class FileSystemComponent {
  protected name: string;

  constructor(name: string) {
    this.name = name;
  }

  public getName(): string {
    return this.name;
  }

  /**
   * Returns the size of the component in bytes.
   */
  public abstract getSize(): number;

  /**
   * Displays the component's structure.
   */
  public abstract display(indent?: string): void;

  // Optional: Default implementations for child management (throw error)
  public add(component: FileSystemComponent): void {
    throw new Error("Cannot add to this component");
  }

  public remove(component: FileSystemComponent): void {
    throw new Error("Cannot remove from this component");
  }

  public getChild(index: number): FileSystemComponent {
    throw new Error("Cannot get child from this component");
  }
}

// --- Leaf Class ---
export class File extends FileSystemComponent {
  private size: number;

  constructor(name: string, size: number) {
    super(name);
    this.size = size;
  }

  public getSize(): number {
    return this.size;
  }

  public display(indent: string = ""): void {
    console.log(`${indent}- ${this.getName()} (${this.getSize()} bytes)`);
  }
}

// --- Composite Class ---
export class Directory extends FileSystemComponent {
  private children: FileSystemComponent[] = [];

  constructor(name: string) {
    super(name);
  }

  public add(component: FileSystemComponent): void {
    this.children.push(component);
  }

  public remove(component: FileSystemComponent): void {
    const index = this.children.indexOf(component);
    if (index > -1) {
      this.children.splice(index, 1);
    } else {
      throw new Error(
        `Component '${component.getName()}' not found in directory '${this.getName()}'`
      );
    }
  }

  public getChild(index: number): FileSystemComponent {
    if (index < 0 || index >= this.children.length) {
      throw new RangeError(
        `Index ${index} out of bounds for directory '${this.getName()}'`
      );
    }
    return this.children[index];
  }

  public getSize(): number {
    let totalSize = 0;
    for (const child of this.children) {
      totalSize += child.getSize();
    }
    return totalSize;
  }

  public display(indent: string = ""): void {
    console.log(`${indent}+ ${this.getName()} (${this.getSize()} bytes total)`);
    for (const child of this.children) {
      child.display(indent + "  ");
    }
  }
}
